modules/photo/src/ccm/ccm.cpp

```diff
@@ -0,0 +1,504 @@
+
+    /** @brief Fitting nonlinear - optimization initial value by white balance.
+        @return the output array, type of Mat
+    */
+    void initialWhiteBalance(void);
+
+    /** @brief Fitting nonlinear-optimization initial value by least square.
+        @param fit if fit is True, return optimalization for rgbl distance function.
+    */
+    void initialLeastSquare(bool fit = false);
+
+    double calcLoss_(Color color);
+    double calcLoss(const Mat ccm_);
+
+    /** @brief Fitting ccm if distance function is associated with CIE Lab color space.
+             see details in https://github.com/opencv/opencv/blob/master/modules/core/include/opencv2/core/optim.hpp
+            Set terminal criteria for solver is possible.
+    */
+    void fitting(void);
+
+    void getColor(Mat& img_, bool islinear = false);
+    void getColor(ColorCheckerType constColor);
+    void getColor(Mat colors_, ColorSpace cs_, Mat colored_);
+    void getColor(Mat colors_, ColorSpace refColorSpace_);
+
+    /** @brief Loss function base on cv::MinProblemSolver::Function.
+             see details in https://github.com/opencv/opencv/blob/master/modules/core/include/opencv2/core/optim.hpp
+    */
+    class LossFunction : public MinProblemSolver::Function
+    {
+    public:
+        ColorCorrectionModel::Impl* ccmLoss;
+        LossFunction(ColorCorrectionModel::Impl* ccm)
+            : ccmLoss(ccm) {};
+
+        /** @brief Reset dims to ccm->shape.
+        */
+        int getDims() const CV_OVERRIDE
+        {
+            return ccmLoss->shape;
+        }
+
+        /** @brief Reset calculation.
+        */
+        double calc(const double* x) const CV_OVERRIDE
+        {
+            Mat ccm_(ccmLoss->shape, 1, CV_64F);
+            for (int i = 0; i < ccmLoss->shape; i++)
+            {
+                ccm_.at<double>(i, 0) = x[i];
+            }
+            ccm_ = ccm_.reshape(0, ccmLoss->shape / 3);
+            return ccmLoss->calcLoss(ccm_);
+        }
+    };
+};
+
+ColorCorrectionModel::Impl::Impl()
+    : cs(*GetCS::getInstance().getRgb(COLOR_SPACE_SRGB))
+    , csEnum(COLOR_SPACE_SRGB)
+    , ccmType(CCM_LINEAR)
+    , distance(DISTANCE_CIE2000)
+    , linearizationType(LINEARIZATION_GAMMA)
+    , weights(Mat())
+    , gamma(2.2)
+    , deg(3)
+    , saturatedThreshold({ 0, 0.98 })
+    , initialMethodType(INITIAL_METHOD_LEAST_SQUARE)
+    , weightsCoeff(0)
+    , maxCount(5000)
+    , epsilon(1.e-4)
+{}
+
+Mat ColorCorrectionModel::Impl::prepare(const Mat& inp)
+{
+    switch (ccmType)
+    {
+    case cv::ccm::CCM_LINEAR:
+        shape = 9;
+        return inp;
+    case cv::ccm::CCM_AFFINE:
+    {
+        shape = 12;
+        Mat arr1 = Mat::ones(inp.size(), CV_64F);
+        Mat arr_out(inp.size(), CV_64FC4);
+        Mat arr_channels[3];
+        split(inp, arr_channels);
+        merge(std::vector<Mat> { arr_channels[0], arr_channels[1], arr_channels[2], arr1 }, arr_out);
```

Comment:
There is `cv::mixChannels` channels to reach the same result faster without extra allocations: https://docs.opencv.org/5.x/d2/de8/group__core__array.html#ga51d768c270a1cdd3497255017c4504be